File: pjos/test/math.c
/*
Copyright 2002 Simon Daniel
email: simon@pjos.org

This file is part of PJOS.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include "interpreter.h"

/*
 * Execute the IADD instruction
 */
void op_iadd() {
    Int value2 = popInt();
    Int value1 = popInt();
    pushInt(value1 + value2);
    pc++;
}

/*
 * Execute the LADD instruction
 */
void op_ladd() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.l = value1.l + value2.l;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the FADD instruction
 */
void op_fadd() {
    Float value2 = popFloat();
    Float value1 = popFloat();
    pushFloat(value1 + value2);
    pc++;
}

/*
 * Execute the DADD instruction
 */
void op_dadd() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.d = value1.d + value2.d;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the ISUB instruction
 */
void op_isub() {
    Int value2 = popInt();
    Int value1 = popInt();
    pushInt(value1 - value2);
    pc++;
}

/*
 * Execute the LSUB instruction
 */
void op_lsub() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.l = value1.l - value2.l;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the FSUB instruction
 */
void op_fsub() {
    Float value2 = popFloat();
    Float value1 = popFloat();
    pushFloat(value1 - value2);
    pc++;
}

/*
 * Execute the DSUB instruction
 */
void op_dsub() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.d = value1.d - value2.d;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the IMUL instruction
 */
void op_imul() {
    Int value2 = popInt();
    Int value1 = popInt();
    pushInt(value1 * value2);
    pc++;
}

/*
 * Execute the LMUL instruction
 */
void op_lmul() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.l = value1.l * value2.l;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the FMUL instruction
 */
void op_fmul() {
    Float value2 = popFloat();
    Float value1 = popFloat();
    pushFloat(value1 * value2);
    pc++;
}

/*
 * Execute the DMUL instruction
 */
void op_dmul() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.d = value1.d * value2.d;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the IDIV instruction
 */
void op_idiv() {
    Int value2 = popInt();
    Int value1 = popInt();
    if (value2 == 0) {
        throwException(CORE_THROW_ARITHMETIC);
        return;
    }
    pushInt(value1 / value2);
    pc++;
}

/*
 * Execute the LDIV instruction
 */
void op_ldiv() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    if (value2.l == 0) {
        throwException(CORE_THROW_ARITHMETIC);
        return;
    }
    value1.l = value1.l / value2.l;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the FDIV instruction
 */
void op_fdiv() {
    Float value2 = popFloat();
    Float value1 = popFloat();
    pushFloat(value1 / value2);
    pc++;
}

/*
 * Execute the DDIV instruction
 */
void op_ddiv() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    value1.d = value1.d / value2.d;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the IREM instruction
 */
void op_irem() {
    Int value2 = popInt();
    Int value1 = popInt();
    if (value2 == 0) {
        throwException(CORE_THROW_ARITHMETIC);
        return;
    }
    pushInt(value1 % value2);
    pc++;
}

/*
 * Execute the LREM instruction
 */
void op_lrem() {
    DoubleWord value2 = popDoubleWord();
    DoubleWord value1 = popDoubleWord();
    if (value2.l == 0) {
        throwException(CORE_THROW_ARITHMETIC);
        return;
    }
    value1.l = value1.l % value2.l;
    pushDoubleWord(value1);
    pc++;
}

/*
 * Execute the FREM instruction
 */
void op_frem() {
    printf("frem not implemented!!!\n");
    exit(1);
    /*
    Float value2 = popFloat();
    Float value1 = popFloat();
    pushFloat(value1 % value2);
    pc++;
    */
}

/*
 * Execute the DREM instruction
 */
void op_drem() {
    printf("drem not implemented!!!\n");
    exit(1);
    /*
    Double value2 = popDouble();
    Double value1 = popDouble();
    pushDouble(value1 % value2);
    pc++;
    */
}

/*
 * Execute the INEG instruction
 */
void op_ineg() {
    Int value = popInt();
    pushInt(-value);
    pc++;
}

/*
 * Execute the LNEG instruction
 */
void op_lneg() {
    DoubleWord dw = popDoubleWord();
    dw.l = -dw.l;
    pushDoubleWord(dw);
    pc++;
}

/*
 * Execute the FNEG instruction
 */
void op_fneg() {
    Float value = popFloat();
    pushFloat(-value);
    pc++;
}

/*
 * Execute the DNEG instruction
 */
void op_dneg() {
    DoubleWord dw = popDoubleWord();
    dw.d = -dw.d;
    pushDoubleWord(dw);
    pc++;
}

/*
 * Execute the IINC instruction
 */
void op_iinc() {
    int index = u8(1);
    Int value = getLocalInt(index);
    setLocalInt(index, value + s8(2), frame);
    pc += 3;
}

File: pjos/org/pjos/emulator/engine/implementation/Math.java
/*
Copyright 2002 Simon Daniel
email: simon@pjos.org

This file is part of PJOS.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
package org.pjos.emulator.engine.implementation;

/**
 * Contains implementations for various maths instructions.
 */
class Math implements Constants {

    /**
     * iadd instruction
     */
    static void iadd() {
        int value2 = Stack.popData();
        int value1 = Stack.popData();
        Stack.pushData(value1 + value2);
        Reg.instruction += 1;
    }

    /**
     * ladd instruction
     */
    static void ladd() {
        long value2 = Stack.popLong();
        long value1 = Stack.popLong();
        Stack.pushLong(value1 + value2);
        Reg.instruction += 1;
    }

    /**
     * fadd instruction
     */
    static void fadd() {
        float value2 = Float.intBitsToFloat(Stack.popData());
        float value1 = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(value1 + value2));
        Reg.instruction += 1;
    }

    /**
     * dadd instruction
     */
    static void dadd() {
        double value2 = Double.longBitsToDouble(Stack.popLong());
        double value1 = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(value1 + value2));
        Reg.instruction += 1;
    }

    /**
     * isub instruction
     */
    static void isub() {
        int value2 = Stack.popData();
        int value1 = Stack.popData();
        Stack.pushData(value1 - value2);
        Reg.instruction += 1;
    }

    /**
     * lsub instruction
     */
    static void lsub() {
        long value2 = Stack.popLong();
        long value1 = Stack.popLong();
        Stack.pushLong(value1 - value2);
        Reg.instruction += 1;
    }

    /**
     * fsub instruction
     */
    static void fsub() {
        float value2 = Float.intBitsToFloat(Stack.popData());
        float value1 = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(value1 - value2));
        Reg.instruction += 1;
    }

    /**
     * dsub instruction
     */
    static void dsub() {
        double value2 = Double.longBitsToDouble(Stack.popLong());
        double value1 = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(value1 - value2));
        Reg.instruction += 1;
    }

    /**
     * imul instruction
     */
    static void imul() {
        int value2 = Stack.popData();
        int value1 = Stack.popData();
        Stack.pushData(value1 * value2);
        Reg.instruction += 1;
    }

    /**
     * lmul instruction
     */
    static void lmul() {
        long value2 = Stack.popLong();
        long value1 = Stack.popLong();
        Stack.pushLong(value1 * value2);
        Reg.instruction += 1;
    }

    /**
     * fmul instruction
     */
    static void fmul() {
        float value2 = Float.intBitsToFloat(Stack.popData());
        float value1 = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(value1 * value2));
        Reg.instruction += 1;
    }

    /**
     * dmul instruction
     */
    static void dmul() {
        double value2 = Double.longBitsToDouble(Stack.popLong());
        double value1 = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(value1 * value2));
        Reg.instruction += 1;
    }

    /**
     * idiv instruction
     */
    static void idiv() {
        int value2 = Stack.popData();
        int value1 = Stack.popData();
        if (value2 == 0) {
            Exceptions.throwException(CORE_THROW_ARITHMETIC);
            return;
        }
        Stack.pushData(value1 / value2);
        Reg.instruction += 1;
    }

    /**
     * ldiv instruction
     */
    static void ldiv() {
        long value2 = Stack.popLong();
        long value1 = Stack.popLong();
        if (value2 == 0) {
            Exceptions.throwException(CORE_THROW_ARITHMETIC);
            return;
        }
        Stack.pushLong(value1 / value2);
        Reg.instruction += 1;
    }

    /**
     * fdiv instruction
     */
    static void fdiv() {
        float value2 = Float.intBitsToFloat(Stack.popData());
        float value1 = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(value1 / value2));
        Reg.instruction += 1;
    }

    /**
     * ddiv instruction
     */
    static void ddiv() {
        double value2 = Double.longBitsToDouble(Stack.popLong());
        double value1 = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(value1 / value2));
        Reg.instruction += 1;
    }

    /**
     * irem instruction
     */
    static void irem() {
        int value2 = Stack.popData();
        int value1 = Stack.popData();
        if (value2 == 0) {
            Exceptions.throwException(CORE_THROW_ARITHMETIC);
            return;
        }
        Stack.pushData(value1 % value2);
        Reg.instruction += 1;
    }

    /**
     * lrem instruction
     */
    static void lrem() {
        long value2 = Stack.popLong();
        long value1 = Stack.popLong();
        if (value2 == 0) {
            Exceptions.throwException(CORE_THROW_ARITHMETIC);
            return;
        }
        Stack.pushLong(value1 % value2);
        Reg.instruction += 1;
    }

    /**
     * frem instruction
     */
    static void frem() {
        float value2 = Float.intBitsToFloat(Stack.popData());
        float value1 = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(value1 % value2));
        Reg.instruction += 1;
    }

    /**
     * drem instruction
     */
    static void drem() {
        double value2 = Double.longBitsToDouble(Stack.popLong());
        double value1 = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(value1 % value2));
        Reg.instruction += 1;
    }

    /**
     * ineg instruction
     */
    static void ineg() {
        Stack.pushData(-Stack.popData());
        Reg.instruction += 1;
    }

    /**
     * lneg instruction
     */
    static void lneg() {
        Stack.pushLong(-Stack.popLong());
        Reg.instruction += 1;
    }

    /**
     * fneg instruction
     */
    static void fneg() {
        float value = Float.intBitsToFloat(Stack.popData());
        Stack.pushData(Float.floatToRawIntBits(-value));
        Reg.instruction += 1;
    }

    /**
     * dneg instruction
     */
    static void dneg() {
        double value = Double.longBitsToDouble(Stack.popLong());
        Stack.pushLong(Double.doubleToRawLongBits(-value));
        Reg.instruction += 1;
    }

}
